File: nodejs/test/e2e/rpc_mcp_lifecycle.e2e.test.ts
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------------------------------------------*/

import { randomUUID } from "node:crypto";
import { dirname, resolve } from "node:path";
import { fileURLToPath } from "node:url";
import { describe, expect, it } from "vitest";
import type { CopilotSession, MCPServerConfig, MCPStdioServerConfig } from "../../src/index.js";
import { approveAll } from "../../src/index.js";
import { createSdkTestContext } from "./harness/sdkTestContext.js";
import { formatError, waitForCondition } from "./harness/sdkTestHelper.js";

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);
const TEST_MCP_SERVER = resolve(__dirname, "../../../test/harness/test-mcp-server.mjs");
const TEST_HARNESS_DIR = dirname(TEST_MCP_SERVER);

describe("Session-scoped MCP lifecycle RPC", async () => {
    const { copilotClient: client } = await createSdkTestContext();

    function createTestMcpServers(...serverNames: string[]): Record<string, MCPServerConfig> {
        return Object.fromEntries(
            serverNames.map((name) => [
                name,
                {
                    type: "local",
                    command: "node",
                    args: [TEST_MCP_SERVER],
                    workingDirectory: TEST_HARNESS_DIR,
                    tools: ["*"],
                } as MCPStdioServerConfig,
            ])
        );
    }

    async function createSessionWithMcp(serverName: string): Promise<CopilotSession> {
        return client.createSession({
            onPermissionRequest: approveAll,
            mcpServers: createTestMcpServers(serverName),
        });
    }

    async function waitForMcpServerStatus(
        session: CopilotSession,
        serverName: string,
        expectedStatus = "connected"
    ): Promise<void> {
        let lastStatus = "<not listed>";
        await waitForCondition(
            async () => {
                const result = await session.rpc.mcp.list();
                const server = result.servers.find((entry) => entry.name === serverName);
                lastStatus = server?.status ?? "<not listed>";
                return server?.status === expectedStatus;
            },
            {
                timeoutMs: 60_000,
                intervalMs: 200,
                timeoutMessage: `${serverName} did not reach ${expectedStatus}; last status was ${lastStatus}`,
            }
        );
    }

    async function waitForMcpRunning(
        session: CopilotSession,
        serverName: string,
        expectedRunning: boolean
    ): Promise<void> {
        await waitForCondition(
            async () =>
                (await session.rpc.mcp.isServerRunning({ serverName })).running === expectedRunning,
            {
                timeoutMs: 60_000,
                intervalMs: 200,
                timeoutMessage: `${serverName} running=${expectedRunning}`,
            }
        );
    }

    function missingName(prefix: string): string {
        return `${prefix}-${randomUUID().replace(/-/g, "")}`;
    }

    function assertNotUnhandledMethod(message: string): void {
        expect(message.toLowerCase()).not.toContain("unhandled method");
    }

    it(
        "should list tools and report running status for connected server",
        { timeout: 120_000 },
        async () => {
            const serverName = "rpc-lifecycle-list-server";
            const session = await createSessionWithMcp(serverName);
            try {
                await waitForMcpServerStatus(session, serverName);

                const tools = await session.rpc.mcp.listTools({ serverName });
                expect(tools.tools.length).toBeGreaterThan(0);
                for (const tool of tools.tools) {
                    expect(tool.name).toBeTruthy();
                }

                expect((await session.rpc.mcp.isServerRunning({ serverName })).running).toBe(true);
                expect(
                    (
                        await session.rpc.mcp.isServerRunning({
                            serverName: missingName("missing"),
                        })
                    ).running
                ).toBe(false);
            } finally {
                await session.disconnect();
            }
        }
    );

    it("should throw when listing tools for unconnected server", { timeout: 120_000 }, async () => {
        const serverName = "rpc-lifecycle-unconnected-host";
        const session = await createSessionWithMcp(serverName);
        try {
            await waitForMcpServerStatus(session, serverName);

            await expect(
                session.rpc.mcp.listTools({ serverName: missingName("missing") })
            ).rejects.toSatisfy((error: unknown) => {
                const message = formatError(error);
                assertNotUnhandledMethod(message);
                expect(message.toLowerCase()).toContain("not connected");
                return true;
            });
        } finally {
            await session.disconnect();
        }
    });

    it("should stop running mcp server", { timeout: 180_000 }, async () => {
        const serverName = "rpc-lifecycle-stop-server";
        const session = await createSessionWithMcp(serverName);
        try {
            await waitForMcpServerStatus(session, serverName);
            expect((await session.rpc.mcp.isServerRunning({ serverName })).running).toBe(true);

            await session.rpc.mcp.stopServer({ serverName });

            await waitForMcpRunning(session, serverName, false);
        } finally {
            await session.disconnect();
        }
    });

    it("should start and restart mcp server", { timeout: 180_000 }, async () => {
        const hostServer = "rpc-lifecycle-host-server";
        const session = await createSessionWithMcp(hostServer);
        try {
            await waitForMcpServerStatus(session, hostServer);

            const startedServer = "rpc-lifecycle-started-server";
            const config = createTestMcpServers(startedServer)[startedServer] as unknown as Record<
                string,
                unknown
            >;

            await session.rpc.mcp.startServer({ serverName: startedServer, config });
            await waitForMcpRunning(session, startedServer, true);

            const tools = await session.rpc.mcp.listTools({ serverName: startedServer });
            expect(tools.tools.length).toBeGreaterThan(0);

            await session.rpc.mcp.restartServer({ serverName: startedServer, config });
            await waitForMcpRunning(session, startedServer, true);
        } finally {
            await session.disconnect();
        }
    });

    it("should register and unregister external mcp client", { timeout: 120_000 }, async () => {
        const hostServer = "rpc-lifecycle-extclient-host";
        const session = await createSessionWithMcp(hostServer);
        try {
            await waitForMcpServerStatus(session, hostServer);

            const externalName = "rpc-lifecycle-external-client";
            expect(
                (await session.rpc.mcp.isServerRunning({ serverName: externalName })).running
            ).toBe(false);

            await session.rpc.mcp.registerExternalClient({
                serverName: externalName,
                client: { id: externalName },
                transport: { kind: "in-process" },
                config: { command: "noop" },
            });
            expect(
                (await session.rpc.mcp.isServerRunning({ serverName: externalName })).running
            ).toBe(true);

            await session.rpc.mcp.unregisterExternalClient({ serverName: externalName });
            expect(
                (await session.rpc.mcp.isServerRunning({ serverName: externalName })).running
            ).toBe(false);
        } finally {
            await session.disconnect();
        }
    });

    it("should reload mcp servers with config", { timeout: 120_000 }, async () => {
        const hostServer = "rpc-lifecycle-reload-host";
        const session = await createSessionWithMcp(hostServer);
        try {
            await waitForMcpServerStatus(session, hostServer);

            const result = await session.rpc.mcp.reloadWithConfig({
                config: {
                    mcpServers: {},
                    disabledServers: [],
                },
            });

            expect(result).toBeDefined();
            expect(result.filteredServers).toEqual([]);
        } finally {
            await session.disconnect();
        }
    });

    it("should configure github mcp server", { timeout: 120_000 }, async () => {
        const hostServer = "rpc-lifecycle-configure-host";
        const session = await createSessionWithMcp(hostServer);
        try {
            await waitForMcpServerStatus(session, hostServer);

            const result = await session.rpc.mcp.configureGitHub({
                authInfo: { type: "api-key" },
            });

            expect(result).toBeDefined();
            expect(result.changed).toBe(false);
        } finally {
            await session.disconnect();
        }
    });

    it(
        "should respond to mcp oauth request without pending request",
        { timeout: 120_000 },
        async () => {
            const hostServer = "rpc-lifecycle-oauth-host";
            const session = await createSessionWithMcp(hostServer);
            try {
                await waitForMcpServerStatus(session, hostServer);

                const result = await session.rpc.mcp.oauth.respond({
                    requestId: missingName("missing"),
                });
                expect(result).toBeDefined();
            } finally {
                await session.disconnect();
            }
        }
    );
});

File: go/internal/e2e/rpc_mcp_lifecycle_e2e_test.go
package e2e

import (
	"strings"
	"testing"
	"time"

	copilot "github.com/github/copilot-sdk/go"
	"github.com/github/copilot-sdk/go/internal/e2e/testharness"
	"github.com/github/copilot-sdk/go/rpc"
)

func TestRpcMcpLifecycle(t *testing.T) {
	ctx := testharness.NewTestContext(t)
	client := ctx.NewClient()
	t.Cleanup(func() { client.ForceStop() })

	t.Run("should_list_tools_and_report_running_status_for_connected_server", func(t *testing.T) {
		ctx.ConfigureForTest(t)
		const serverName = "rpc-lifecycle-list-server"
		session := createPortedSession(t, client, &copilot.SessionConfig{MCPServers: testMCPServers(t, serverName)})
		defer session.Disconnect()
		waitForPortedMCPServerStatus(t, session, serverName, rpc.MCPServerStatusConnected)

		tools, err := session.RPC.MCP.ListTools(t.Context(), &rpc.MCPListToolsRequest{ServerName: serverName})
		if err != nil {
			t.Fatalf("MCP.ListTools failed: %v", err)
		}
		if len(tools.Tools) == 0 {
			t.Fatal("Expected connected MCP server to expose at least one tool")
		}
		for _, tool := range tools.Tools {
			if strings.TrimSpace(tool.Name) == "" {
				t.Fatalf("Expected non-empty MCP tool name, got %+v", tool)
			}
		}

		running, err := session.RPC.MCP.IsServerRunning(t.Context(), &rpc.MCPIsServerRunningRequest{ServerName: serverName})
		if err != nil {
			t.Fatalf("MCP.IsServerRunning(%s) failed: %v", serverName, err)
		}
		if !running.Running {
			t.Fatalf("Expected %s to be running", serverName)
		}
		missing, err := session.RPC.MCP.IsServerRunning(t.Context(), &rpc.MCPIsServerRunningRequest{ServerName: "missing-" + randomHex(t)})
		if err != nil {
			t.Fatalf("MCP.IsServerRunning(missing) failed: %v", err)
		}
		if missing.Running {
			t.Fatal("Expected missing MCP server not to be running")
		}
	})

	t.Run("should_throw_when_listing_tools_for_unconnected_server", func(t *testing.T) {
		ctx.ConfigureForTest(t)
		const serverName = "rpc-lifecycle-unconnected-host"
		session := createPortedSession(t, client, &copilot.SessionConfig{MCPServers: testMCPServers(t, serverName)})
		defer session.Disconnect()
		waitForPortedMCPServerStatus(t, session, serverName, rpc.MCPServerStatusConnected)

		_, err := session.RPC.MCP.ListTools(t.Context(), &rpc.MCPListToolsRequest{ServerName: "missing-" + randomHex(t)})
		if err == nil {
			t.Fatal("Expected MCP.ListTools for an unconnected server to fail")
		}
		message := err.Error()
		assertPortedNoUnhandledMethod(t, message)
		assertPortedContainsFold(t, message, "not connected")
	})

	t.Run("should_stop_running_mcp_server", func(t *testing.T) {
		ctx.ConfigureForTest(t)
		const serverName = "rpc-lifecycle-stop-server"
		session := createPortedSession(t, client, &copilot.SessionConfig{MCPServers: testMCPServers(t, serverName)})
		defer session.Disconnect()
		waitForPortedMCPServerStatus(t, session, serverName, rpc.MCPServerStatusConnected)
		waitForPortedMCPRunning(t, session, serverName, true)

		if _, err := session.RPC.MCP.StopServer(t.Context(), &rpc.MCPStopServerRequest{ServerName: serverName}); err != nil {
			t.Fatalf("MCP.StopServer failed: %v", err)
		}
		waitForPortedMCPRunning(t, session, serverName, false)
	})

	t.Run("should_start_and_restart_mcp_server", func(t *testing.T) {
		ctx.ConfigureForTest(t)
		const hostServer = "rpc-lifecycle-host-server"
		session := createPortedSession(t, client, &copilot.SessionConfig{MCPServers: testMCPServers(t, hostServer)})
		defer session.Disconnect()
		waitForPortedMCPServerStatus(t, session, hostServer, rpc.MCPServerStatusConnected)

		const startedServer = "rpc-lifecycle-started-server"
		config := testMCPServers(t, startedServer)[startedServer]
		if _, err := session.RPC.MCP.StartServer(t.Context(), &rpc.MCPStartServerRequest{ServerName: startedServer, Config: config}); err != nil {
			t.Fatalf("MCP.StartServer failed: %v", err)
		}
		waitForPortedMCPRunning(t, session, startedServer, true)

		tools, err := session.RPC.MCP.ListTools(t.Context(), &rpc.MCPListToolsRequest{ServerName: startedServer})
		if err != nil {
			t.Fatalf("MCP.ListTools(started) failed: %v", err)
		}
		if len(tools.Tools) == 0 {
			t.Fatal("Expected started MCP server to expose tools")
		}

		if _, err := session.RPC.MCP.RestartServer(t.Context(), &rpc.MCPRestartServerRequest{ServerName: startedServer, Config: config}); err != nil {
			t.Fatalf("MCP.RestartServer failed: %v", err)
		}
		waitForPortedMCPRunning(t, session, startedServer, true)
	})

	t.Run("should_register_and_unregister_external_mcp_client", func(t *testing.T) {
		ctx.ConfigureForTest(t)
		const hostServer = "rpc-lifecycle-extclient-host"
		session := createPortedSession(t, client, &copilot.SessionConfig{MCPServers: testMCPServers(t, hostServer)})
		defer session.Disconnect()
		waitForPortedMCPServerStatus(t, session, hostServer, rpc.MCPServerStatusConnected)

		const externalName = "rpc-lifecycle-external-client"
		initial, err := session.RPC.MCP.IsServerRunning(t.Context(), &rpc.MCPIsServerRunningRequest{ServerName: externalName})
		if err != nil {
			t.Fatalf("MCP.IsServerRunning(initial external) failed: %v", err)
		}
		if initial.Running {
			t.Fatal("Expected external client to start as not running")
		}

		if _, err := session.RPC.MCP.RegisterExternalClient(t.Context(), &rpc.MCPRegisterExternalClientRequest{
			ServerName: externalName,
			Client:     map[string]any{"id": externalName},
			Transport:  map[string]any{"kind": "in-process"},
			Config:     map[string]any{"command": "noop"},
		}); err != nil {
			t.Fatalf("MCP.RegisterExternalClient failed: %v", err)
		}
		waitForPortedMCPRunning(t, session, externalName, true)

		if _, err := session.RPC.MCP.UnregisterExternalClient(t.Context(), &rpc.MCPUnregisterExternalClientRequest{ServerName: externalName}); err != nil {
			t.Fatalf("MCP.UnregisterExternalClient failed: %v", err)
		}
		waitForPortedMCPRunning(t, session, externalName, false)
	})

	t.Run("should_reload_mcp_servers_with_config", func(t *testing.T) {
		ctx.ConfigureForTest(t)
		const hostServer = "rpc-lifecycle-reload-host"
		session := createPortedSession(t, client, &copilot.SessionConfig{MCPServers: testMCPServers(t, hostServer)})
		defer session.Disconnect()
		waitForPortedMCPServerStatus(t, session, hostServer, rpc.MCPServerStatusConnected)

		result, err := session.RPC.MCP.ReloadWithConfig(t.Context(), &rpc.MCPReloadWithConfigRequest{Config: map[string]any{
			"mcpServers":      map[string]any{},
			"disabledServers": []string{},
		}})
		if err != nil {
			t.Fatalf("MCP.ReloadWithConfig failed: %v", err)
		}
		if result == nil {
			t.Fatal("Expected non-nil reload result")
		}
		if result.FilteredServers == nil {
			t.Fatal("Expected non-nil FilteredServers")
		}
		if len(result.FilteredServers) != 0 {
			t.Fatalf("Expected no filtered servers, got %+v", result.FilteredServers)
		}
	})

	t.Run("should_configure_github_mcp_server", func(t *testing.T) {
		ctx.ConfigureForTest(t)
		const hostServer = "rpc-lifecycle-configure-host"
		session := createPortedSession(t, client, &copilot.SessionConfig{MCPServers: testMCPServers(t, hostServer)})
		defer session.Disconnect()
		waitForPortedMCPServerStatus(t, session, hostServer, rpc.MCPServerStatusConnected)

		result, err := session.RPC.MCP.ConfigureGitHub(t.Context(), &rpc.MCPConfigureGitHubRequest{AuthInfo: map[string]any{"type": "api-key"}})
		if err != nil {
			t.Fatalf("MCP.ConfigureGitHub failed: %v", err)
		}
		if result == nil {
			t.Fatal("Expected non-nil configure result")
		}
		if result.Changed {
			t.Fatal("Expected Changed=false")
		}
	})

	t.Run("should_respond_to_mcp_oauth_request_without_pending_request", func(t *testing.T) {
		ctx.ConfigureForTest(t)
		const hostServer = "rpc-lifecycle-oauth-host"
		session := createPortedSession(t, client, &copilot.SessionConfig{MCPServers: testMCPServers(t, hostServer)})
		defer session.Disconnect()
		waitForPortedMCPServerStatus(t, session, hostServer, rpc.MCPServerStatusConnected)

		result, err := session.RPC.MCP.Oauth().Respond(t.Context(), &rpc.MCPOauthRespondRequest{RequestID: "missing-" + randomHex(t)})
		if err != nil {
			t.Fatalf("MCP.Oauth.Respond failed: %v", err)
		}
		if result == nil {
			t.Fatal("Expected non-nil OAuth respond result")
		}
	})
}

func waitForPortedMCPServerStatus(t *testing.T, session *copilot.Session, serverName string, expectedStatus rpc.MCPServerStatus) {
	t.Helper()
	waitForRPCCondition(t, 60*time.Second, serverName+" reaching "+string(expectedStatus), func() (bool, error) {
		result, err := session.RPC.MCP.List(t.Context())
		if err != nil {
			return false, err
		}
		for _, server := range result.Servers {
			if server.Name == serverName {
				return server.Status == expectedStatus, nil
			}
		}
		return false, nil
	})
}

func waitForPortedMCPRunning(t *testing.T, session *copilot.Session, serverName string, expectedRunning bool) {
	t.Helper()
	waitForRPCCondition(t, 60*time.Second, serverName+" running state", func() (bool, error) {
		result, err := session.RPC.MCP.IsServerRunning(t.Context(), &rpc.MCPIsServerRunningRequest{ServerName: serverName})
		if err != nil {
			return false, err
		}
		return result.Running == expectedRunning, nil
	})
}
